database/merkledb/tree.go

```diff
@@ -0,0 +1,281 @@
+// a subset of database content with a particular key prefix starting at a
+// specified key.
+func (t *Tree) NewIteratorWithStartAndPrefix(start, prefix []byte) database.Iterator {
+	return NewIteratorWithStartAndPrefix(t, start, prefix)
+}
+
+func (t *Tree) Stat(property string) (string, error) {
+	return "nil", nil
+}
+
+func (t *Tree) Compact(start []byte, limit []byte) error {
+	if t.isClosed() != nil {
+		return t.isClosed()
+	}
+	return nil
+}
+
+func (t *Tree) Close() error {
+	if t.isClosed() != nil {
+		return t.isClosed()
+	}
+
+	t.closed = true
+	return t.persistence.db.(*versiondb.Database).GetDatabase().Close()
+}
+
+// NewMemoryTree returns a new instance of the Tree with a in-memoryDB
+func NewMemoryTree() *Tree {
+	return NewTree(memdb.New())
+}
+
+// NewLevelTree returns a new instance of the Tree with a in-memoryDB
+func NewLevelTree(file string) *Tree {
+	db, err := leveldb.New(file, 0, 0, 0)
+	if err != nil {
+		panic(err)
+	}
+	return NewTree(db)
+}
+
+// NewTree returns a new instance of the Tree
+func NewTree(db database.Database) *Tree {
+	persistence, _ := NewPersistence(db)
+	return &Tree{
+		closed:      false,
+		persistence: persistence,
+	}
+}
+
+func (t *Tree) Root() ([]byte, error) {
+	rootNode, err := t.persistence.GetRootNode()
+	if err != nil {
+		return nil, err
+	}
+	return rootNode.GetHash(), nil
+}
+
+func (t *Tree) Get(key []byte) ([]byte, error) {
+	if t.isClosed() != nil {
+		return nil, t.isClosed()
+	}
+
+	rootNode, err := t.persistence.GetRootNode()
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := t.findNode(BytesToKey(key), rootNode)
+	if err != nil {
+		return nil, err
+	}
+	if node == nil {
+		return nil, database.ErrNotFound
+	}
+	if _, ok := node.(*EmptyNode); ok {
+		return nil, database.ErrNotFound
+	}
+
+	return node.Value(), nil
+}
+
+// Put travels the tree and finds the node to insert the LeafNode
+func (t *Tree) Put(key []byte, value []byte) error {
+	return t.persistence.Commit(t.put(key, value))
+}
+
+func (t *Tree) Delete(key []byte) error {
+	return t.persistence.Commit(t.del(key))
+}
+
+func (t *Tree) findNode(key Key, node Node) (Node, error) {
+
+	if node == nil {
+		return nil, nil
+	}
+
+	switch node.(type) {
+	case *EmptyNode:
+		return node, nil
+	case *LeafNode:
+		return node, nil
+	}
+
+	nodeChild, err := node.GetChild(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return t.findNode(key, nodeChild)
+}
+
+func (t *Tree) PrintTree() {
+	rootNode, _ := t.persistence.GetRootNode()
+	rootNode.Print()
+}
+
+func (t *Tree) fetchNextNode(prefix Key, start Key, key Key, node Node) (Node, error) {
+	if node == nil || t.closed {
+		return nil, database.ErrClosed
+	}
+
+	switch node.(type) {
+	case *EmptyNode:
+		return nil, nil
+	case *LeafNode:
+		return node, nil
+	}
+
+	nextNode, err := node.GetNextNode(prefix, start, key)
+	if err != nil {
+		return nil, err
+	}
+	return t.fetchNextNode(prefix, start, key, nextNode)
+}
+
+func (t *Tree) isClosed() error {
```

Comment:
It's because I'm lazy haha.
I thought this function might change - some places we check the open/close with different strategies like checking for nil.

I just abstracted that out. Instead of having to change in multiple places I just use that func.

Reply:
Okay, but I don't see how using this function makes live any easier. You still need an if statement to check if this function returns an error... In all the cases this function is used I think it's cleaner (and less work) to just inline this function. For example from `tree.go#*Tree.Has`:
```
if t.isClosed() != nil {
	return false, t.isClosed()
}
```
turns into:
```
if t.closed {
	return false, database.ErrClosed
}
```
I'm not sure if the compiler is smart enough to do this translation on the fly, If I were to guess I'd say it is because this function is probably getting inlined... But regardless of any potential speedup, this just seem kludgy.

Reply:
Just the case if t.closed needed to change to nil or something else.

That said, I don't think it will change much from now on, and that is much cleaner.